twister: in case of samples, test on integration platforms if defined

nashif opened the pull request:
Add option to force integration mode on a defined list of tests, for
example tests for sample that are identified with 'sample.'.

A sample per definition shall be tested on documented and supported
platforms listed in the sample documentation. Samples shall not be used
as tests to verify functionality of a feature on all available plaforms
in Zephyr.

To still allow testing on platforms not listed in the doc, and when such
platforms are covered by the provided filter, we should still be able to
build/run the tests on those platforms (not directly listed in
integration_platforms).

We detect a sample by its test identifier, i.e., it starts with 'sample.'.

nashif committed:
twister: config: test on integration plaforms if defined

Add option to force integration mode on a defined list of tests, for
example tests for sample that are identified with 'sample.'.

A sample per definition shall be tested on documented and supported
platforms listed in the sample documentation. Samples shall not be used
as tests to verify functionality of a feature on all available plaforms
in Zephyr.

To still allow testing on platforms not listed in the doc, and when such
platforms are covered by the provided filter, we should still be able to
build/run the tests on those platforms (not directly listed in
integration_platforms).

We detect a sample by its test identifier, i.e., it starts with 'sample.'.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

hakehuang commented:
@nashif thanks a lot to add the feature, so what if the sample.yaml has `platform_allow`? the platform has to be in platform_allow? or we are moving `platform_allow` to `integration_platforms`?

nashif commented:
> @nashif thanks a lot to add the feature, so what if the sample.yaml has `platform_allow`? the platform has to be in platform_allow? or we are moving `platform_allow` to `integration_platforms`?

platform_allow and all other behavior stays the same, platform_allow is still a valid filter like before. `integration_platforms` should continue to only have a minimal set of platforms tested in CI and should not be a copy of `platform_allow`

golowanow commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -788,7 +788,12 @@ def apply_filters(self, **kwargs):
-                        platform_scope = platforms + integration_platforms
+                        # if integration mode or when dealing with samples, only run on
+                        # integration platforms if integration_platforms is set.
+                        if 'sample.' in ts.id:
```
```suggestion
                        if ts.id.startswith('sample.'):
```

gchwier reviewed (dismissed):
`apply_filters`  is one of the methods that should be refactored, as its logic is quite complex. In the past, I tried to understand how platform selection works and created some diagrams. These diagrams may not be up-to-date, but they might still be useful to someone.

Pre-selected platforms:

![image2023-3-23_14-39-6](https://github.com/user-attachments/assets/86706c38-437a-4a40-b38a-9dbcbb40f3c2)

Platform scope (per testsuite):

![image2023-3-23_14-40-41](https://github.com/user-attachments/assets/89b2ebe1-065e-45c1-b3dc-640b902aa250)

PerMac reviewed (changes requested):
TL;DR: IMO such logic should be in our CI workflow where twister is called and not in twister itself. If we don't want to build samples for all platforms, why we even use `--all` parameter in our CI in the first place?

I see the need for getting better control over what's built and why. I was thinking a bit about this change and argumentation behind and I am still puzzled. This solution adds some "if logic" in already complex function in twister, overriding other "decision making" parameters, to mitigate a fact, that we don't want to build samples on all platforms in the projects CI. I am worried that this makes it even harder to figure out how scope selection should work and validation of the correct behavior. 
We have "--all" flag in twister that suppose to select all available platforms into the scope, but now, you'll have to add disclaimer "*doesn't affect what has "sample" in name". Isn't it counter intuitive? There are other ways of calling twister that would also require clarifications. 

What if for whatever reason someone would like to verify if a sample executes correctly on all "default testing platforms" (i.e. multiple qemus and simulators)? 
`./scripts/twister -T samples/hello_world/ -v` with this PR won't give you execution on 27 platforms but just native_sim. 
Now a user will have to add dozens/hundreds of platforms with `-p plaform X` to twister call if they want to use any other grouping than "integration_platforms". This can affect downstreams as well. I understand the motivation, that it should help downstreams to not be blocked by platform_allow, but I think it causes other issues I tried to point out.  

 How I would tackle the issue: 
first find where we build samples on all the platforms instead of selected few. For now i believe the main workflow for twister is https://github.com/zephyrproject-rtos/zephyr/blob/main/.github/workflows/twister.yaml. It has 3 triggers:
- weekly, where the `--all` is used. If we don't want to build all samples there, let's just change the twister command for this trigger. `-T tests` would overwrite default `tests and samples` leaving only tests. Samples on integration_platforms were already verified on other triggers. If we want to include samples in the same run for the sake of completeness, we can split twister call into 2. one for tests with `--all` and one for samples with `--integration`.
- on-push: can be done the same way as above: split twister call into `-T tests` and `-T samples --integration`. Tests will get "default" scope, samples "integration" one.
- on-PR: Test_plan.py is in use. If we modify a given sample than only that sample will be verified but on all (not descoped) platforms. I don't think it is a burden for the CI. I would leave it as it is. If we modify multiple files and/or test_plan.py cannot dynamically limit the scope, the --integration flag is already in place, hence only integration_platforms will be selected regardless if it is test or sample.

PerMac commented:
https://github.com/zephyrproject-rtos/zephyr/pull/86154

nashif committed:
ci: twister: use ci specific test configuration

Use a CI specific test configuration that forces samples to be built in
integration mode.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@PerMac  please revisit
